src/translator/processed_request_sentence.h

```diff
@@ -0,0 +1,68 @@
+#ifndef SRC_BERGAMOT_PROCESSED_REQUEST_SENTENCE_H_
+#define SRC_BERGAMOT_PROCESSED_REQUEST_SENTENCE_H_
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "translator/beam_search.h"
+
+namespace marian {
+namespace bergamot {
+
+/// History is marian object with plenty of `shared_ptr`s lying around, which makes keeping a lid on memory hard for
+/// cache, due to copies being shallow rather than deep. We therefore process marian's lazy History management changing
+/// to something eager, keeping only what we want and units accounted for. Each of these corresponds to a
+/// `RequestSentence` and retains the minimum information required to reconstruct a full `Response` if cached.
+///
+/// L4 requires byte-streams, which allows for it to keep tight control over "memory" we need to have accounting for
+/// cache-size based on memory/bytes as well for any arbitrary serializable into bytestream data-types.
+///
+/// An empty() method is provided which relays if no words are present. This empty type is also treated as  an
+/// incomplete ProcessedRequestSentence during cache operations.
+class ProcessedRequestSentence {
```

Comment:
This whole class can be replaced with just always having the byte representation and accessing it directly, rather than serialization and deserialization.

Reply:
```
class ProcessedRequestSentence {
  public:
    float sentenceScore() const { return reinterpret_cast<const float*>(base_); }
    // Could have an object with begin and end methods that looks like a vector but is actually just a range.  
    const marian::Word *words() const;
    const float *wordScores() const;
    const float *alignmentMatrix() const;
    std::size_t length() const;
  private:
    void *base_;
};
```

Reply:
Since the general concern here is conversions and copies...

I'm on board with the idea here. There are, however, a bunch of ownership concerns. L4 owns `Key` and `Value`. I have to `memcpy` `Value = (data*, size)` (as a binary-sequence and have `ProcessRequestSentence` own the copied entity) because the cache (value) entry can be evicted potentially in a background thread once fetch is complete and the element is RAII Object is out of scope. I don't want to leak L4 all the way up. 

There are still (unavoidable) copies when we chop down `marian::History` (freshly computed, not cached) to be thinner and then written onto something flat and sequential (as L4 requires continuous byte-sequence for storing Value). 

With words put on to this flat-storage:


```cpp 
    const marian::Word *words() const;
	const marian::Word *words_begin_;
```

Marian decoding API (https://github.com/browsermt/marian-dev/blob/62bac858bfd37060beb707d12eb9711649ea4cf6/src/data/sentencepiece_vocab.cpp#L259) works with `marian::Words & == vector<marian::Word> &`. The standard vector has to own the underlying storage I can't find a way to override and provide a pseudo const vector that uses the sequential binary flat thing as storage.

I'm looking at adding the `(words*, size)` decoding API (which involve edits to marian-dev) to avoid a copy here. An alternative is a copy onto vector at decoding, which I will do to check things stay intact before spilling over to more repositories (for now at least).

Please let know if I'm missing something or some understanding is wrong.

For now, I have created binary storage within `ProcessedRequestSentence` which is the flat thing asked for and forbid copy on storage, which in-turn forbids copy on `ProcessedRequestSentence` which should eliminate unnecessary copies. Conversions however still exist.

Reply:
Regarding the marian API with Words, it's already a bergamot special API anyway.  The best practice here is to replace it with accepting an iterator range.  Or at least a pointer range.

Reply:
https://github.com/browsermt/marian-dev/issues/54